respy/method_of_simulated_moments.py

```diff
@@ -0,0 +1,316 @@
+    empirical_moments : list
+        Contains the empirical moments calculated for the observed data. Each item in
+        the list constitutes a set of moments saved to a pandas.DataFrame or
+        pandas.Series. Index of pandas.DataFrames can be of type MultiIndex, but columns
+        cannot.
+    weighting_matrix : numpy.ndarray
+        Square matrix of dimension (NxN) with N denoting the number of
+        empirical_moments. Used to weight squared moment errors.
+    return_scalar : bool
+        Indicates whether to return moment error vector (False) or weighted square
+        product of moment error vector (True).
+
+    Returns
+    -------
+    out : pandas.Series or float
+        Scalar or moment error vector depending on value of return_scalar.
+
+    """
+    empirical_moments = copy.deepcopy(empirical_moments)
+
+    df = simulate(params)
+    simulated_moments = [func(df) for func in calc_moments]
+
+    simulated_moments = [
+        sim_mom.reindex_like(emp_mom)
+        for emp_mom, sim_mom in zip(empirical_moments, simulated_moments)
+    ]
+
+    simulated_moments = [
+        func(mom) for mom, func in zip(simulated_moments, replace_nans)
+    ]
+
+    flat_empirical_moments = _flatten_index(empirical_moments)
+    flat_simulated_moments = _flatten_index(simulated_moments)
+
+    moment_errors = flat_empirical_moments - flat_simulated_moments
+
+    # Return moment errors as indexed DataFrame or calculate weighted square product of
+    # moment errors depending on return_scalar.
+    if return_scalar is False:
+        out = moment_errors
+    else:
+        out = moment_errors.T @ weighting_matrix @ moment_errors
+
+    return out
+
+
+def get_diag_weighting_matrix(empirical_moments, weights=None):
+    """Create a diagonal weighting matrix from weights.
+
+    Parameters
+    ----------
+    empirical_moments : pandas.DataFrame or pandas.Series or dict or list
+        Contains the empirical moments calculated for the observed data. Moments should
+        be saved to pandas.DataFrame or pandas.Series that can either be passed to the
+        function directly or as items of a list or dictionary.
+    weights : pandas.DataFrame or pandas.Series or dict or list
+        Contains weights (usually variances) of empirical moments. Must match structure
+        of empirical_moments i.e. if empirical_moments is a list of pandas.DataFrames,
+        weights be list of pandas.DataFrames as well where each DataFrame entry contains
+        the weight for the corresponding moment in empirical_moments.
+
+    Returns
+    -------
+    numpy.ndarray
+        Array contains a diagonal weighting matrix.
+
+    """
+    weights = copy.deepcopy(weights)
+    empirical_moments = copy.deepcopy(empirical_moments)
+    empirical_moments = _harmonize_input(empirical_moments)
+
+    # Use identity matrix if no weights are specified.
+    if weights is None:
+        weights = copy.deepcopy(empirical_moments)
```

Comment:
I just noticed it's probably easier to just fill in the 1's for the flattend empirical moments since this removes the for-loop. Or how would you implement a list comprehension here?